src/main/java/net/coderbot/iris/pipeline/TransformPatcher.java

```diff
@@ -0,0 +1,512 @@
+
+					@Override
+					public void enterArrayAccessExpression(ArrayAccessExpressionContext ctx) {
+						ParseTreeMatch match = textureMatrixPattern.match(ctx);
+						if (match.succeeded()) {
+							replaceNode(ctx, "mat4(1.0)", GLSLParser::expression);
+						}
+					}
+				});
+			}
+		};
+
+		/**
+		 * Implements BuiltinUniformReplacementTransformer and does a little more. Note
+		 * that the main walk phase uses the fact that the order of invocation is
+		 * enterMemberAccessExpression, enterMultiplicativeExpression,
+		 * enterArrayAccessExpression in the targeted expression.
+		 */
+		LifecycleUser<Parameters> replaceBuiltinUniforms = new Transformation<Parameters>() {
+			static final String lightmapCoordsExpression = "a_LightCoord";
+			static final String irisLightmapTexMat = "iris_LightmapTextureMatrix";
+			static final String texCoordFallbackReplacement = "vec4(" + lightmapCoordsExpression + " * 255.0, 0.0, 1.0)";
+
+			boolean needsLightmapTexMatInjection;
+
+			@Override
+			public void resetState() {
+				needsLightmapTexMatInjection = false;
+			}
+
+			{
+				// make sure the lightmap coords expression doesn't exist in the code yet
+				addEndDependent(new SearchTerminals<Parameters>()
+						.addTarget(new WrapThrowTargetImpl<>(lightmapCoordsExpression))
+						.addTarget(new WrapThrowTargetImpl<>(irisLightmapTexMat)));
+
+				// find accesses to gl_TextureMatrix[1] or gl_TextureMatrix[2] in combination
+				// with gl_MultiTexCoord1 or gl_MultiTexCoord2 and replace them with
+				// lightmapCoordsExpression or a vector wrapper depending on the context
+				chainDependent(new WalkPhase<Parameters>() {
+					ParseTreePattern accessPattern;
+					ParseTreePattern bareMultPattern;
+					ParseTreePattern extraPattern;
+					ParseTreePattern textureMatrixPattern;
+
+					private void checkPatternMatch(ParseTreePattern pattern, ExtendedContext ctx,
+							Consumer<ParseTreeMatch> action) {
+						ParseTreeMatch match = pattern.match(ctx);
+						if (match.succeeded()) {
+							String texCoord = match.get("texCoord").getText();
+							String texMatrixIndex = match.get("texMatrixIndex").getText();
+							if ((texCoord.equals("gl_MultiTexCoord1") || texCoord.equals("gl_MultiTexCoord2"))
+									&& (texMatrixIndex.equals("1") || texMatrixIndex.equals("2"))) {
+								action.accept(match);
+							}
+						}
+					}
+
+					@Override
+					public void init() {
+						accessPattern = compilePattern(
+								"(gl_TextureMatrix[<texMatrixIndex:expression>] * <texCoord:IDENTIFIER>).<member:IDENTIFIER>",
+								GLSLParser.RULE_expression);
+						bareMultPattern = compilePattern(
+								"gl_TextureMatrix[<texMatrixIndex:expression>] * <texCoord:IDENTIFIER>",
+								GLSLParser.RULE_expression);
+						extraPattern = compilePattern("<texCoord:IDENTIFIER>.xy / 255.0", GLSLParser.RULE_expression);
+						textureMatrixPattern = compilePattern("gl_TextureMatrix[<texMatrixIndex:expression>]",
+								GLSLParser.RULE_expression);
```

Comment:
I don't have any comments here other than to say that this system looks extremely powerful to use!

Reply:
Thank you for your acknowledgement, it really means a lot to me. I'll continue to work on getting trunk up to date to the latest merged changes and working through your comments.